Add Request.is_authenticated and is_authenticated predicate

merwok opened the pull request:
Fixes #1602

merwok commented on src/pyramid/interfaces.py:
```diff
@@ -233,6 +241,10 @@ def encode_content(encoding='gzip', lazy=False):
 
     headers = Attribute(""" The headers in a dictionary-like object """)
 
+    is_authenticated = Attribute(
```
@mmerickel @stevepiercy Does it follow the naming conventions to use a verb here for a property?  It feels a bit weird to me, but there is no reason to make it a method (it relies on `authenticated_identity` which is itself a property that calls many methods).  `authenticated` alone doesn’t seem so good either.

stevepiercy replied:
Are you struggling with the question, "What is authenticated by `is_authenticated`?" If so, then `thing_is_authenticated`. If not, can you elaborate on what you want to convey to the developer with the name of the property?

mmerickel replied:
The naming `is_authenticated` is what I was expecting.

merwok commented on src/pyramid/security.py:
```diff
@@ -244,6 +244,11 @@ def authenticated_userid(self):
+    @property
+    def is_authenticated(self):
+        """Return True if a user is authenticated for this request."""
+        return self.authenticated_identity is not None
```
To do: confirm that this is the right attribute to check.  I have forgotten if identity/userid depend on each other, and if yes which one is the primary one.

mmerickel replied:
this is ok for now

merwok commented on tests/test_config/test_predicates.py:
```diff
@@ -38,6 +39,19 @@ def test_ordering_xhr_and_request_method_trump_only_containment(self):
     def test_ordering_number_of_predicates(self):
         from pyramid.config.predicates import predvalseq
 
+        order0, _, _ = self._callFUT(
```
I named this object 0 to avoid having to rename everyting and mess up the git blame on lines 139-148.

mmerickel replied:
seems good to me

stevepiercy commented:
@merwok fyi, thanks for the offer to update What's New, but it is easier to do a single copy-pasta job at release time. See https://github.com/Pylons/pyramid/blob/master/RELEASING.txt#L49-L58 for process.

merwok commented:
I keep wanting to add a note somewhere about this, but I don’t know if it’s appropriate:

This configuration will not work because the two view configs have the same specificity:

```python
@view_config(is_authenticated=True)
def home_members(request):
    ...

@view_config(is_authenticated=False)
def home_login(request):
    ....
```

The right one is add is_authenticated to one of the functions and leave the other one without the param.  It may be obvious to Pyramid devs, and the example in the docs is correct (https://github.com/Pylons/pyramid/pull/3598/files#diff-c0f6b28a439e308028cc66c9f0ac7aee), but it did trip me up for a moment on a project.

mmerickel commented:
> This configuration will not work because the two view configs have the same specificity:

They have the same view weighting but because they are properly mutually exclusive, the correct one will be selected for each request. This configuration will work.

There could definitely be issues with effective_principals because you could setup configurations that were not mutually exclusive.

merwok commented:
Well that’s great!  I think now that my error was probably that my code was doing `return request.authenticated_userid is not None` without comparing to the boolean value stored as `self.val` :man_facepalming: :laughing:

stevepiercy commented on src/pyramid/predicates.py:
```diff
@@ -276,6 +276,19 @@ def __call__(self, context, request):
+    phash = text
+
+    def __call__(self, context, request):
+        return bool(request.is_authenticated) is self.val
```
Is `request.is_authenticated` already a boolean? If so, then is wrapping with `bool(...)` necessary? Or should `self.val` be wrapped, i.e., `bool(self.val)`?

merwok replied:
The code was `request.is_authenticated == self.val` before, but I changed it to match XhrPredicate.  I don’t understand the reason (maybe because `request.is_xhr` is not controlled by Pyramid but webob?), just thought the style should match.  I can easily revert.

mmerickel replied:
either way is fine to me but if I were writing it myself I would've used `==`

stevepiercy commented on docs/narr/viewconfig.rst:
```diff
@@ -494,6 +494,16 @@ configured view.
 
+``is_authenticated``
+
+  This value, if specified, should be either ``True`` or ``False``.  If it is
```
```suggestion
  This value, if specified, must be either ``True`` or ``False``.  If it is
```

This might need clarification pending outcome of comment https://github.com/Pylons/pyramid/pull/3598/files#diff-dcbd22655007a5407eaee38bdc2e1cf5R289

merwok replied:
`should` is taken from the documentation of other predicates.

stevepiercy replied:
Given that, should we document what happens if the value is neither `True` nor `False`, or leave it open for the developer to find out?

merwok replied:
I have changed to `must`.  Is more documentation needed?

stevepiercy replied:
`must` is more definitive, so I think this is good to go.

stevepiercy commented on docs/narr/viewconfig.rst:
```diff
@@ -494,6 +494,16 @@ configured view.
+  specified and is ``True``, the request must be for an authenticated user,
+  as determined by the :term:`security policy` in use.  If it is specified and
+  ``False``, the associated view callable will match only if the request does
+  not have an authenticated user.
```
```
If it is specified and its value is ``True``, only a request from an
authenticated user, as determined by the :term:`security policy`
in use, will satisfy this predicate.

If it is specified and its value is ``False``, only a request from
a user who is not authenticated will satisfy this predicate.
```

The `is_authenticated` predicate is applicable to both routes and views, and this version makes it copy-pasteable into route config. I like less maintenance!

Also I don't think it is correct to say that a "view callable will match", but maybe "predicate will match" or "predicate is satisfied". I took a shot at it, but maybe @mmerickel can wordsmith it better. Whatever is correct in Pyramid terminology.

The sentences should be merged into one paragraph. I put them separately for easier comprehension.

merwok replied:
Oops, that should have been `the view config will match`!  (not correct to say `the view callable will be invoked`, there are other predicates)

merwok replied:
I have kept the doc as it was (not copy-pastable between route and view doc), to follow the existing style.

stevepiercy replied:
Yeah, the predicate docs are suboptimal. There's an [open issue](https://github.com/Pylons/pyramid/issues/3597) that describes the current state, identifies the problems, and offers suggestions to improve the situation. I'd appreciate input on the best direction forward, then I'll create a PR for it.

merwok replied:
I’m not sure about my wording `the view config will match`, it seems a little imprecise.
(`the route will match` feels correct, if a little dry — maybe `route definition`?)

I agree now that `the predicate will match` is simple, correct and reusable so I will edit the new docs.

stevepiercy commented on src/pyramid/config/routes.py:
```diff
@@ -268,6 +268,17 @@ def add_route(
+          authenticated user.
+
+          .. versionadded:: 2.0
+
```
Perhaps update according to comment on `docs/narr/viewconfig.rst `.

stevepiercy replied:
For accuracy, it would be good to account for multiple predicates.

stevepiercy commented on src/pyramid/config/views.py:
```diff
@@ -712,6 +712,16 @@ def wrapper(context, request):
+          will match only if the request does not have an authenticated user.
+
+          ..versionadded:: 2.0
+
```
Perhaps update according to comment on `docs/narr/viewconfig.rst `.

stevepiercy replied:
For accuracy, it would be good to account for multiple predicates.

stevepiercy commented on src/pyramid/interfaces.py:
```diff
@@ -113,6 +113,16 @@ def app_iter_range(start, stop):
+    authenticated_identity = Attribute(
+        """An object representing the authenticated user, as determined by
+        the security policy in use.  The object's class and meaning is defined
+        by the security policy.  Will be None for unauthenticated requests."""
```
```suggestion
        by the security policy.  Will be ``None`` for unauthenticated requests."""
```

stevepiercy commented on src/pyramid/interfaces.py:
```diff
@@ -113,6 +113,16 @@ def app_iter_range(start, stop):
+    )
+
+    authenticated_userid = Attribute(
+        """A string to identify the authenticated user, or None."""
```
```suggestion
        """A string to identify the authenticated user or ``None``."""
```

stevepiercy commented on src/pyramid/interfaces.py:
```diff
@@ -233,6 +243,11 @@ def encode_content(encoding='gzip', lazy=False):
     headers = Attribute(""" The headers in a dictionary-like object """)
 
+    is_authenticated = Attribute(
+        """A boolean indicated whether the request has an authenticated
```
```suggestion
        """A boolean indicating whether the request has an authenticated
```

stevepiercy commented on src/pyramid/interfaces.py:
```diff
@@ -233,6 +243,11 @@ def encode_content(encoding='gzip', lazy=False):
 
+    is_authenticated = Attribute(
+        """A boolean indicated whether the request has an authenticated
+        user (determined by the security policy in use)."""
```
```suggestion
        user, as determined by the security policy in use."""
```

For consistency.

stevepiercy commented on src/pyramid/security.py:
```diff
@@ -244,6 +244,11 @@ def authenticated_userid(self):
 
+    @property
+    def is_authenticated(self):
+        """Return True if a user is authenticated for this request."""
```
```suggestion
        """Return ``True`` if a user is authenticated for this request."""
```

merwok commented on docs/narr/viewconfig.rst:
```diff
@@ -505,6 +515,8 @@ configured view.
 
   .. versionadded:: 1.4a4
 
+  .. deprecated:: TODO add
```
I think this should say `2.0`, right?

mmerickel replied:
Correct

mmerickel commented on CHANGES.rst:
```diff
@@ -22,10 +22,14 @@ Features
   - ``pyramid.config.Configurator.set_security_policy``.
   - ``pyramid.interfaces.ISecurityPolicy``
   - ``pyramid.request.Request.authenticated_identity``.
+  - ``pyramid.request.Request.authenticated_userid``
```
`authenticated_userid` - this one is not new

mmerickel commented on CHANGES.rst:
```diff
@@ -22,10 +22,14 @@ Features
+  - ``pyramid.request.Request.is_authenticated``
   - ``pyramid.authentication.SessionAuthenticationHelper``
   - ``pyramid.authorization.ACLHelper``
+  - config predicate ``is_authenticated=True/False``
```
```
- `is_authenticated=True/False` route and view predicates
```

merwok commented:
Oh also let me validate the whole thing with a tiny app!  (can only trust unit tests so far)

merwok commented:
I have added a bunch of routes and views to a minimal app and it looks good.

<details>
<summary>Python</summary>

```python
    # tests for is_authenticated
    config.add_route("members-only", "/cool-zone", is_authenticated=True)
    config.add_view(hello_member, route_name="members-only")

    config.add_route("home", "/home")
    config.add_route("welcome", "/home", is_authenticated=False)
    config.add_view(hello_member, route_name="home")
    config.add_view(hello_stranger, route_name="welcome")

    config.add_route("history", "/history")
    config.add_view(hello_stranger, route_name="history")
    config.add_view(hello_member, route_name="history", is_authenticated=True)

    config.add_route("profile", "/profile")
    config.add_view(hello_stranger, route_name="profile", is_authenticated=False)
    config.add_view(hello_member, route_name="profile", is_authenticated=True)
```
</details>


<details>
<summary>proutes</summary>

```
Name                         Pattern                              View                                                                     Method    
----                         -------                              ----                                                                     ------    
hello                        /                                    helloapp.hello_world                                                     *         
route_to_num                 /{num}                               <unknown>                                                                *         
ymd2                         /{year:\d+}/{month:\d+}/{day:\d+}    helloapp.hello_world                                                     *         
y                            /{year}                              <unknown>                                                                *         
ym                           /{year}/{month}                      <unknown>                                                                *         
ymd                          /{year}/{month}/{day}                helloapp.hello_world                                                     *         
members-only                 /cool-zone                           helloapp.hello_member                                                    *         
home                         /home                                helloapp.hello_member                                                    *         
welcome                      /home                                helloapp.hello_stranger                                                  *         
history                      /history                             pyramid.config.views.<pyramid.config.views.MultiView object at 0x...>    *         
profile                      /profile                             pyramid.config.views.<pyramid.config.views.MultiView object at 0x...>    *         
debugtoolbar                 /_debug_toolbar/*subpath             <unknown>                                                                *         
__/_debug_toolbar/static/    /_debug_toolbar/static/*subpath      pyramid_debugtoolbar:static/                                             *         
```   

</details>

<details>
<summary>pviews</summary>

```
URL = /home

    context: <pyramid.traversal.DefaultRootFactory object at 0x...>
    view name: home

    Route:
    ------
    route name: home
    route pattern: /home
    route path: /home
    subpath: 

        View:
        -----
        helloapp.hello_member.

    Route:
    ------
    route name: welcome
    route pattern: /home
    route path: /home
    subpath: 
    route predicates (is_authenticated = False)

        View:
        -----
        helloapp.hello_stranger.

URL = /cool-zone

    context: <pyramid.traversal.DefaultRootFactory object at 0x...>
    view name: cool-zone

    Route:
    ------
    route name: members-only
    route pattern: /cool-zone
    route path: /cool-zone
    subpath: 
    route predicates (is_authenticated = True)

        View:
        -----
        helloapp.hello_member.


URL = /profile

    context: <pyramid.traversal.DefaultRootFactory object at 0x...>
    view name: profile

    Route:
    ------
    route name: profile
    route pattern: /profile
    route path: /profile
    subpath: 

        View:
        -----
        helloapp.hello_stranger
        view predicates (is_authenticated = False)

        View:
        -----
        helloapp.hello_member
        view predicates (is_authenticated = True)


URL = /history

    context: <pyramid.traversal.DefaultRootFactory object at 0x...>
    view name: history

    Route:
    ------
    route name: history
    route pattern: /history
    route path: /history
    subpath: 

        View:
        -----
        helloapp.hello_member
        view predicates (is_authenticated = True)

        View:
        -----
        helloapp.hello_stranger
```

</details>

I also deleted the original IsAuthenticated predicate in my project, installed this Pyramid branch and my tests all pass.

stevepiercy commented on docs/narr/viewconfig.rst:
```diff
@@ -494,6 +494,16 @@ configured view.
+  specified and is ``True``, the request must be for an authenticated user,
+  as determined by the :term:`security policy` in use.  If it is specified and
+  ``False``, the associated view callable will be invoked only if the request
+  does not have an authenticated user.
```
If docs are incorrect, copying them is not the best option. In this case, it looks like the original author did not account for multiple predicates on a view or route config. I'd prefer to update the copy to make it correct. Would you please do so per [your comment](https://github.com/Pylons/pyramid/pull/3598#discussion_r449010244)?

Ideally the original docs from which the concept was copied would also get corrected, or at least an issue opened or amended. See https://github.com/Pylons/pyramid/issues/3597

merwok commented:
@mmerickel this is ready!

bertjwregeer commented:
\o/ this is great, thank you @merwok!
